File: api/bubble-data.py
import json
from urllib.parse import parse_qs, urlparse

import pandas as pd

from api._lib.blob_io import load_df
from api._lib.handler_utils import BaseHandler, find_clicks_col, get_processed_df
from api._lib.core.utils import norm


class handler(BaseHandler):
    def do_GET(self):
        try:
            qs = parse_qs(urlparse(self.path).query)
            raw_df = load_df(qs.get("rawDfUrl", [""])[0])
            n = max(1, int(qs.get("n", ["500"])[0]))
            group_col = qs.get("groupCol", [None])[0]

            state_raw = qs.get("state", [None])[0]
            state = json.loads(state_raw) if state_raw else {}
            processed = get_processed_df(raw_df, state) if state else raw_df.copy()

            clicks_col = find_clicks_col(processed)
            title_col = next(
                (c for c in processed.columns if norm(c) == norm("title")),
                processed.columns[0] if len(processed.columns) else None,
            )
            nums = (
                pd.to_numeric(processed[clicks_col].astype(str).str.extract(r"([\d.]+)")[0], errors="coerce").fillna(0)
                if clicks_col else pd.Series(0.0, index=processed.index)
            )
            bubble_df = processed.copy()
            bubble_df["_clicks"] = nums
            bubble_df = bubble_df.sort_values("_clicks", ascending=False).head(n).reset_index(drop=True)

            items = [
                {
                    "title": str(row.get(title_col, "")) if title_col else "",
                    "clicks": float(row.get("_clicks", 0) or 0),
                    "category": str(row.get(group_col, "")).strip() if group_col and group_col in bubble_df.columns else "",
                }
                for _, row in bubble_df.iterrows()
            ]
            group_candidates = [
                c for c in processed.columns
                if 2 <= int(processed[c].astype(str).str.strip().replace("", pd.NA).dropna().nunique()) <= 50
                and c not in {"_clicks", title_col}
            ]

            self._send_json(200, {
                "items": items,
                "total": len(raw_df),
                "groupCandidates": group_candidates,
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

File: api/bubble-data.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import { sendError } from '../api/_lib/handlerUtils.js'
import { loadRows } from '../api/_lib/sessionStore.js'
import { getProcessedRows } from '../api/_lib/pipeline.js'
import { findClicksCol } from '../api/_lib/core/feedProcessor.js'
import { norm, toNumber, isBlank } from '../api/_lib/core/utils.js'

const EXCLUDED_COLS = new Set(['price_currency', 'sale price_currency'])

export default async function handler(req: VercelRequest, res: VercelResponse) {
  if (req.method !== 'POST') return res.status(405).json({ error: 'Method not allowed' })

  try {
    const { rawDfUrl, n = 500, groupCol, state: params = {} } = req.body as {
      rawDfUrl: string
      n?: number
      groupCol?: string
      state?: Record<string, unknown>
    }

    const raw = await loadRows(rawDfUrl)
    const processed = await getProcessedRows(raw, params)

    const cols = Object.keys(processed[0] ?? {})
    const clicksCol = findClicksCol(cols)
    const priceCol = !clicksCol ? (cols.find(c => norm(c) === 'price') ?? null) : null
    const valueCol = clicksCol ?? priceCol
    const valueLabel = clicksCol ? 'clicks' : priceCol ? 'price' : ''
    const titleCol = cols.find(c => norm(c) === 'title') ?? null
    const candidates = cols.filter(c => !EXCLUDED_COLS.has(c))

    const topN = Number(n) || 500

    const items = processed
      .map(row => ({
        title: titleCol ? String(row[titleCol] ?? '') : '',
        value: valueCol ? (toNumber(row[valueCol]) || 0) : 0,
        category: groupCol ? String(row[groupCol] ?? '') : '',
      }))
      .filter(item => !isBlank(item.title))
      .sort((a, b) => b.value - a.value)
      .slice(0, topN)

    res.status(200).json({
      items,
      total: processed.length,
      groupCandidates: candidates,
      valueLabel,
    })
  } catch (err) {
    sendError(res, err)
  }
}
